File: src/main/java/io/github/saneea/feature/random/RandomBytes.java
package io.github.saneea.feature.random;

import java.io.IOException;
import java.io.OutputStream;
import java.util.Random;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.Option;

import io.github.saneea.Feature;
import io.github.saneea.FeatureContext;

public class RandomBytes implements Feature, Feature.CLI, Feature.Out.Bin.Stream {

	private static final String SIZE = "size";
	private static final String SEED = "seed";

	private OutputStream out;

	private CommandLine commandLine;

	@Override
	public String getShortDescription() {
		return "generates random binary data";
	}

	@Override
	public void run(FeatureContext context) throws IOException {

		long count = Long.parseLong(commandLine.getOptionValue(SIZE));

		String seed = commandLine.getOptionValue(SEED);

		Random r = seed == null //
				? new Random()
				: new Random(Long.parseLong(seed));

		byte[] bytes = new byte[1];

		for (int i = 0; i < count; ++i) {
			r.nextBytes(bytes);
			out.write(bytes);
		}
	}

	@Override
	public void setOut(OutputStream out) {
		this.out = out;
	}

	@Override
	public void setCommandLine(CommandLine commandLine) {
		this.commandLine = commandLine;
	}

	@Override
	public Option[] getOptions() {
		return new Option[] { //

				Option//
						.builder("s")//
						.longOpt(SIZE)//
						.hasArg(true)//
						.argName("bytes count")//
						.required(true)//
						.desc("size of generated data in bytes")//
						.build(), //

				Option//
						.builder()//
						.longOpt(SEED)//
						.hasArg(true)//
						.argName("seed number")//
						.required(false)//
						.desc("seed for random generator")//
						.build()//
		};
	}

}

File: src/main/java/io/github/saneea/dvh/feature/random/RandomBytes.kt
package io.github.saneea.dvh.feature.random

import io.github.saneea.dvh.Feature
import io.github.saneea.dvh.Feature.Meta
import io.github.saneea.dvh.FeatureContext
import org.apache.commons.cli.CommandLine
import org.apache.commons.cli.Option
import java.io.OutputStream
import java.util.Random

class RandomBytes :
    Feature,
    Feature.CLI,
    Feature.CLI.Options,
    Feature.Out.Bin.Stream {

    override lateinit var outBinStream: OutputStream
    override lateinit var commandLine: CommandLine

    override fun meta(context: FeatureContext) =
        Meta("generates random binary data")

    override fun run(context: FeatureContext) {
        val count = commandLine.getOptionValue(SIZE).toLong()
        val seed: String? = commandLine.getOptionValue(SEED)

        val r = if (seed == null)
            Random()
        else
            Random(seed.toLong())

        val bytes = ByteArray(1)

        for (i in 0 until count) {
            r.nextBytes(bytes)
            outBinStream.write(bytes)
        }
    }

    override val options: Array<Option>
        get() {
            return arrayOf(
                Option
                    .builder("s")
                    .longOpt(SIZE)
                    .hasArg(true)
                    .argName("bytes count")
                    .required(true)
                    .desc("size of generated data in bytes")
                    .build(),
                Option
                    .builder()
                    .longOpt(SEED)
                    .hasArg(true)
                    .argName("seed number")
                    .required(false)
                    .desc("seed for random generator")
                    .build()
            )
        }

    companion object {
        private const val SIZE = "size"
        private const val SEED = "seed"
    }
}